src/main/kotlin/com/demonwav/mcdev/i18n/intentions/SortTranslationsIntention.kt

```diff
@@ -0,0 +1,128 @@
+                        scope = Scope.PROJECT,
+                        domain = I18nElementFactory.getResourceDomain(psiFile.virtualFile)
+                    )
+                    val indices = defaults.mapIndexed { i, prop -> prop.key to i }.toMap()
+                    assembleElements(project, it.sortedBy { indices[it.key] ?: Int.MAX_VALUE })
+                }
+            }
+        }
+
+        psiFile.runWriteAction {
+            for (elem in psiFile.children) {
+                elem.delete()
+            }
+            for (elem in sorted) {
+                psiFile.add(elem)
+            }
+        }
+    }
+
+    private fun sortByTemplate(project: Project, template: Template, entries: List<I18nEntry>): List<PsiElement> {
+        val result = mutableListOf<PsiElement>()
+        val tmp = entries.toMutableList()
+        for (elem in template.elements) {
+            when (elem) {
+                is Comment -> {
+                    result.add(I18nElementFactory.createComment(project, elem.text))
+                    result.add(I18nElementFactory.createLineEnding(project))
+                }
+                EmptyLine -> result.add(I18nElementFactory.createLineEnding(project))
+                is Key -> {
+                    val toWrite = tmp.filter { elem.matcher.matches(it.key) }
+                    result.addAll(assembleElements(project, toWrite.sortedWith(ascendingComparator)))
+                    tmp.removeAll(toWrite)
+                }
+            }
+        }
+        if (tmp.isNotEmpty()) {
+            result.addAll(assembleElements(project, tmp.sortedWith(ascendingComparator)))
+        }
+        return result
+    }
+
+    private fun assembleElements(project: Project, elements: List<I18nEntry>): List<PsiElement> {
+        val result = mutableListOf<PsiElement>()
+        val withComments = elements.associate { it to gatherComments(it) }
+        for ((entry, comments) in withComments) {
+            for (comment in comments) {
+                result.add(I18nElementFactory.createComment(project, comment))
+                result.add(I18nElementFactory.createLineEnding(project))
+            }
+            result.add(I18nElementFactory.createEntry(project, entry.key, entry.value))
+            result.add(I18nElementFactory.createLineEnding(project))
+        }
+        return result
+    }
+
+    private tailrec fun gatherComments(element: PsiElement, acc: List<String> = emptyList(), depth: Int = 0): List<String> {
+        if (keepComments != 0 && depth >= keepComments) {
+            return acc
+        }
+        val prev = element.prevSibling ?: return acc
+        if (prev.node.elementType != I18nTypes.LINE_ENDING) {
+            return acc
+        }
+        val prevLine = prev.prevSibling ?: return acc
+        if (prevLine.node.elementType != I18nTypes.COMMENT) {
+            return acc
+        }
+        return gatherComments(prevLine, listOf(prevLine.text.substring(1).trim()) + acc, depth + 1)
```

Comment:
Instead of doing list copies every iteration, `acc` should just be a `MutableList`. The return type of the method can still be `List`.